src/main/java/gim/model/exercise/ExerciseHashMap.java

```diff
@@ -20,8 +23,40 @@ public class ExerciseHashMap {
 
     private final HashMap<Name, ArrayList<Exercise>> exerciseHashMap;
 
+    private final ArrayList<Observer> observerArrayList;
+
+    /**
+     * Constructs a {@code ExerciseHashMap}.
+     */
     public ExerciseHashMap() {
         exerciseHashMap = new HashMap<>();
+        observerArrayList = new ArrayList<Observer>();
+    }
+
+    /**
+     * Constructs a {@code ExerciseHashMap} with a given ArrayList of Observers
+     * @param arrL
+     */
+    public ExerciseHashMap(ArrayList<Observer> arrL) {
+        this.exerciseHashMap = new HashMap<>();
+        this.observerArrayList = arrL;
+    }
+
+    /**
+     * Notifies observers who have 'subscribed' whenever there are changes in the state of the ExerciseHashMap
+     */
+    public void notifyObservers() {
+        for (Observer o: observerArrayList) {
+            o.update();
+        }
+    }
+
+    /**
+     * Allows an Observer object to 'subscribe' to changes in the state of the ExerciseHashMap.
+     * @param o Observer object
+     */
+    public void addUI(Observer o) {
```

Comment:
Should the naming of this method be `addUi` instead?

Reply:
Necessary changes made, thanks.